checker/manual/inference.tex

```diff
@@ -0,0 +1,248 @@
+make the program type-check.
+These qualifiers are inserted into the source code or the
+class file.  They can be viewed and adjusted by the programmer, and can
+be used by tools such as the type-checker.
+
+Advantages of this variety of type inference include:
+\begin{itemize}
+\item
+  The inference may be more precise by taking account of the entire program
+  rather than just reasoning one method at a time.
+\item
+  The program source code contains documentation in the form of type
+  qualifiers, which can aid programmer understanding and may make
+  type-checking warnings more comprehensible.
+\end{itemize}
+
+
+\subsection{Type inference tools\label{type-inference-tools}}
+
+This section lists tools that take a program and output a set of
+annotations for it.
+
+
+\subsubsection{Type inference for specific type systems\label{type-inference-tools-specialized}}
+
+This section lists type inference tools that work for only a single type system.
+Also see Section~\ref{type-inference-tools-general}, for type inference
+tools that work for any type system.
+
+Section~\ref{nullness-inference} lists several tools that infer
+annotations for the Nullness Checker.
+
+Section~\ref{javari-inference} lists a tool that infers
+annotations for the Javari Checker, which detects mutation errors.
+
+If you run the Checker Framework with the \<-AsuggestPureMethods>
+command-line option, it will suggest methods that can be marked as
+\<@SideEffectFree>, \<@Deterministic>, or \<@Pure>; see
+Section~\ref{type-refinement-purity}.
+
+
+\subsubsection{Type inference for any type system\label{type-inference-tools-general}}
+
+This section lists type inference tools that work for any type system.
+Also see Section~\ref{type-inference-tools-specialized}, for type inference
+tools that work for only a single type system.
+
+By supplying the \<-AinferSignatures> command-line option,
+any type-checker can infer annotations. See Section~\ref{signature-inference}.
+
+\href{https://github.com/reprogrammer/cascade/}{Cascade}~\cite{VakilianPEJ2014}
+is an Eclipse plugin that implements interactive type qualifier inference.
+Cascade is interactive rather than fully-automated:  it makes it easier for
+a developer to insert annotations.
+Cascade starts with an unannotated program and runs a type-checker.  For each
+warning it suggests multiple fixes, the developer chooses a fix, and
+Cascade applies it.  Cascade works with any checker built on the Checker
+Framework.
+You can find installation instructions and a video tutorial at \url{https://github.com/reprogrammer/cascade}.
+
+
+\subsection{Signature inference (interprocedural type refinement)\label{signature-inference}}
+
+Elsewhere in this manual, type refinement (Section~\ref{type-refinement})
+has been described as working only within a method.  In fact, it can also
+infer types for fields, method parameters, and method return types, and it
+can insert these inferred types into your program.
+
+To do so, run the script
```

Comment:
It's important to note that the CHECKERFRAMEWORK environment variable must be set and that insert-annotations-to-source must be on the path.